docs(proposal): Introduce a custom resource to represent a (trusted) registry (#76)
aquasecurity/starboard: proposals/trusted_registry_crd.md

# Introduce a custom resource to represent a (trusted) registry

## TOC

- [Abstract](#abstract)
- [Problem statement](#problem-statement)
- [Solution](#solutions)
- [Use cases](#use-cases)
  - [Starboard Security Operator supervising trusted registries](#starboard-security-operator-supervising-trusted-registries)
  - [Repository webhooks integrated with Starboard](#repository-webhooks-integrated-with-starboard)

## Abstract

Common requirements for running secure Kubernetes workloads are:

- Pull container images only from trusted registries and/or repositories. For example, one might allow pulling

>   images from private Harbor repository accessible at https://core.harbor.domain/library/nginx:1.16, but deny pull
>   requests to the core.harbor.domain/library/redis:5 or https://docker.io/library/nginx:1.16 repositories.

Reviewer: I might be leaping to wrong assumptions here but I think the tag is irrelevant? Suggested changes for clarification (but maybe I have the wrong end of the stick?)

```suggestion
  images from private Harbor repository accessible at https://core.harbor.domain/library/nginx, but deny pull
  requests to https://core.harbor.domain/library/redis (unapproved repo `redis`) or https://docker.io/library/nginx (unapproved registry `docker.io`).
```

> - Statically scan container images for critical, high or medium vulnerabilities. In general, these rules are configurable.

Reviewer: ```suggestion
- Statically scan container images, and implement rules to deny workloads that include critical, high or medium vulnerabilities. In general, these rules are configurable (for example, what level of severity will cause an image to be denied).
```

  `kubectl run redis-pod --image core.harbor.domain/library/redis:5`.

## Problem statement

- Starboard CLI and Starboard Security Operator are able to scan workloads already deployed in a cluster. Hence, an 
  admission controller cannot use vulnerability reports for compliance checks.
- The `vulnerabilities.aquasecurity.github.io` resources are scoped to the same namespace as the underlying workload.
  However, for admission controllers we might want to introduce cluster-scoped resource to represent vulnerability
  report associated with a given container image digest, for example `registryvulnerabilities.aquasecurity.github.io`.
- Container images are not first-class citizens in Kubernetes-native world. There’s no built-in resource that represents
  a container image. Also, there’s no built-in representation of a (trusted) registry.
- An admission controller has certain performance requirements. Usually we cannot trigger a scanner when the admission
  controller is invoked for a given Kubernetes API request. Most likely it will take way too long to perform such
  ad hoc scans.

## Solutions

> ### Define registries.aquasecurity.github.io resource

Reviewer: Nit - should this be `registry` in the singular (like `pod` or `node`)? I think this differs from vulnerabilities because we have many vulns in one instance

Author: Usually the names that you get from `k api-resources --api-group ""` are plural. Also fully qualified resource names must be plural in CRD descriptors:

```
k explain crd.spec.names.plural
KIND:     CustomResourceDefinition
VERSION:  apiextensions.k8s.io/v1

FIELD:    plural <string>

DESCRIPTION:
     plural is the plural name of the resource to serve. The custom resources
     are served under `/apis/<group>/<version>/.../<plural>`. Must match the
     name of the CustomResourceDefinition (in the form
     `<names.plural>.<group>`). Must be all lowercase
```

First off, introduce a custom resource to represent a (trusted) container registry. For example, we can represent
a private Harbor registry with the following Kubernetes object:

```yaml
apiVersion: aquasecurity.github.io/v1alpha1

> kind: Registry

Reviewer 2: Is this resource required to create Registry Vulnerability? I think it is unnecessary for some users to specify trusted registries. They might just want to audit vulnerabilities in the image they're trying to deploy.

Author: I agree. I think we should be able to create vulnerability reports for the specified image digest and store it at the cluster level without requiring the registry resource.

metadata:
  name: my-harbor-registry
spec:
  name: My Harbor Registry
  description: |
    That's the only registry authorized to run production workloads.
  server: core.harbor.domain
  imagePullSecrets:
  - name: my-harbor-registry-credentials
  insecure: false
  trustedRepositories:

>   - name: library/nginx:1.16

Reviewer: Do you think we need to restrict down to the level of approved / not-approved tags? I was imagining more at the repo level. Or even just to the registry level. 

Registries - only allow images from registries that have a CRD instance in place 
Repositories - do they have to have a trustedRepositories entry? Or maybe the default is to allow everything from all repos in a trusted registry? (I am not totally convinced we need the granularity of repository-level control for the first version...) 

Edit: on further reading I think you are suggesting using this as the list of repos that the Starboard operator knows to scan. I wonder if this is conflating two different things. We could have a CRD for trusted registries, and a Starboard config map for a list of repos to be scanned? I am not sure if this is better, just raising it as an option. 

Having a CRD for trusted registries would allow for a super-simple admission controller that denies images that don't come from a trusted registry.

Author: - The first use case for such CRD that came to my mind was an admission controller that required a container image to be already scanned. For example, the admission controller could reject a deployment if:
  - the container image hasn't been scanned
  - the container image has already been scanned but has a critical vulnerability
- In other words, my assumption was that we do reject deployments by default, unless they're scanned
- The other use case for the registries CRD is to specify the desired state of which container images we want to be scanned. Whereas the actual state is represented by the corresponding vulnerability reports. So the desired state could be reconciled by the Starboard Security Operator if the integration via webhooks was not possible

Author: And I agree, we should think more about the granularity of such allow list, whether it's registry, repository, or tag/digest level.

  - name: library/alpine
  - namePattern: trusted/**
status:
  trustedRepositoryStatuses:
  - name: library/nginx
    scanned: true
    lastScanTime: 2020-01-01T00:00:00Z
  - name: library/alpine
    scanned: false
    lastScanTime: null
```

> The `Registry` entity has the `spec` property, which allows specifying trusted repositories. The `status` filed might

Reviewer: ```suggestion
The `Registry` entity has the `spec` property, which allows specifying trusted repositories. The `status` field might
```

contain scan status for individual repositories. In the example above, you can see that the `library/nginx` repository

> has already been scanned, whereas the `library/alpine` repository has not.

Reviewer: What if the repo is part-scanned? Some image versions might be scanned, others might not...

Author: I think that eventually an admission controller might be checking existence of the registryvulnerabilities reports for the given deployment request. The status at the registry level would be just a summary / current status of the reconciliation loop.

### Define registryvulnerabilities.aquasecurity.github.io resource

Define the `RegistryVulnerability` entity to represent a vulnerability report for the specified image digest:

```yaml
apiVersion: aquasecurity.github.io/v1alpha1
kind: RegistryVulnerability
metadata:
  name: ddbb8bfb-040a-42a1-a051-d5aff77a53e7
  labels:
    starboard.repository.name: library/nginx
    starboard.repository.tag: 1.16
    starboard.repository.digest: sha256.d20aa6d1cae56fd17cd458f4807e0de462caf2336f0b70b5eeb69fcaaf30dd9c
    starboard.resource.kind: Registry
    starboard.resource.name: my-harbor-registry
report:
  scanner:
    name: Trivy
    vendor: Aqua Security
    version: 0.9.1
  registry:
    url: https://core.harbor.domain
  artifact:
      repository: "library/nginx"
      digest: "sha256:d20aa6d1cae56fd17cd458f4807e0de462caf2336f0b70b5eeb69fcaaf30dd9c"
      tag: "1.16"
      mimeType: "application/vnd.docker.distribution.manifest.v2+json"
  summary:
    criticalCount: 0
    highCount: 0
    lowCount: 92
    mediumCount: 27
    noneCount: 0
    unknownCount: 3
  vulnerabilities:

>   - id: CVE-2020-1879

Reviewer 2: Did you just omit details such as a package name and CVSS? They are necessary so that a complicated policy can be applied in Admission Controller.

Author: We'll have all the details as we have for `vulnerabilities.aquasecurity.github.io`. For example, https://github.com/aquasecurity/starboard/blob/master/kube/examples/vulnerabilities.yaml#L30 I've update the sample snippet.
